File: app/src/main/java/com/jiujiu/mvvmTemplate/ui/products/details/ProductDetailViewModel.java
package com.jiujiu.mvvmTemplate.ui.products.details;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import com.jiujiu.mvvmTemplate.data.DataManager;
import com.jiujiu.mvvmTemplate.data.model.Product;
import com.jiujiu.mvvmTemplate.ui.base.BaseViewModel;

import javax.inject.Inject;

import io.reactivex.schedulers.Schedulers;

public class ProductDetailViewModel extends BaseViewModel {

    private static final String TAG = "ProductDetailViewModel";

    private MutableLiveData<Product> mProductLiveData = new MutableLiveData<>();

    @Inject
    public ProductDetailViewModel(DataManager dataManager) {
        super(dataManager);
    }

    public void loadProductById(Long productId) {
        getCompositeDisposable().add(
                this.getDataManager().getProductById(productId)
                        .subscribeOn(Schedulers.io())
                        .subscribe(product -> {
                            if (product != null)
                                mProductLiveData.postValue(product);
                        }, throwable -> {
                            throwable.printStackTrace();
                        })
        );
    }

    public LiveData<Product> getProductLiveData() {
        return mProductLiveData;
    }
}

File: app/src/main/java/com/jiujiu/mvvmTemplate/ui/products/details/ProductDetailViewModel.kt
package com.jiujiu.mvvmTemplate.ui.products.details

import androidx.lifecycle.LiveData
import androidx.lifecycle.MediatorLiveData
import androidx.lifecycle.viewModelScope
import com.jiujiu.mvvmTemplate.data.DataManager
import com.jiujiu.mvvmTemplate.data.model.Product
import com.jiujiu.mvvmTemplate.ui.base.BaseViewModel
import kotlinx.coroutines.launch
import javax.inject.Inject

class ProductDetailViewModel @Inject constructor(
        dataManager: DataManager
) : BaseViewModel(dataManager) {

    private var mProductLiveData: MediatorLiveData<Product> = MediatorLiveData()

    val productLiveData: LiveData<Product>
        get() = mProductLiveData

    fun loadProductById(productId: Long?) {
        viewModelScope.launch {
            mProductLiveData.addSource(dataManager.getProductById(productId)) { mProductLiveData.postValue(it) }
        }
    }

}
